infrastructure/devicepathresolver/iscsi_device_path_resolver.go

```diff
@@ -0,0 +1,184 @@
+package devicepathresolver
+
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	boshopeniscsi "github.com/cloudfoundry/bosh-agent/platform/openiscsi"
+	boshsettings "github.com/cloudfoundry/bosh-agent/settings"
+	boshdirs "github.com/cloudfoundry/bosh-agent/settings/directories"
+	bosherr "github.com/cloudfoundry/bosh-utils/errors"
+	boshlog "github.com/cloudfoundry/bosh-utils/logger"
+	boshsys "github.com/cloudfoundry/bosh-utils/system"
+)
+
+// iscsiDevicePathResolver resolves device path by performing Open-iscsi discovery
+type iscsiDevicePathResolver struct {
+	diskWaitTimeout time.Duration
+	runner          boshsys.CmdRunner
+	openiscsi       boshopeniscsi.OpenIscsi
+	fs              boshsys.FileSystem
+	dirProvider     boshdirs.Provider
+	logTag          string
+	logger          boshlog.Logger
+}
+
+func NewIscsiDevicePathResolver(
+	diskWaitTimeout time.Duration,
+	runner boshsys.CmdRunner,
+	openiscsi boshopeniscsi.OpenIscsi,
+	fs boshsys.FileSystem,
+	dirProvider boshdirs.Provider,
+	logger boshlog.Logger,
+) DevicePathResolver {
+	return iscsiDevicePathResolver{
+		diskWaitTimeout: diskWaitTimeout,
+		runner:          runner,
+		openiscsi:       openiscsi,
+		fs:              fs,
+		dirProvider:     dirProvider,
+		logTag:          "iscsiResolver",
+		logger:          logger,
+	}
+}
+
+func (ispr iscsiDevicePathResolver) GetRealDevicePath(diskSettings boshsettings.DiskSettings) (string, bool, error) {
+	var lastDiskID string
+
+	if diskSettings.InitiatorName == "" {
+		return "", false, bosherr.Errorf("iSCSI InitiatorName is not set")
+	}
+
+	if diskSettings.Username == "" {
+		return "", false, bosherr.Errorf("iSCSI Username is not set")
+	}
+
+	if diskSettings.Password == "" {
+		return "", false, bosherr.Errorf("iSCSI Password is not set")
+	}
+
+	if diskSettings.Target == "" {
+		return "", false, bosherr.Errorf("iSCSI Iface Ipaddress is not set")
```

Comment:
Since the property is called "target" please add that to error message... in other words make sure the message is referring to the name of the property in manifest. Otherwise user could be confused.